### PR TITLE
Bug fixes and enhancements to date/time formatters

### PR DESCRIPTION
**Pull request by rich-iannone**
This PR provides a bugfix, more flexibility, improved documentation, and more testing for the date/time formatting functions (`fmt_date()`, `fmt_time()`, and `fmt_datetime()`). Now `Date` and `POSIXct` columns are allowed to be formatted with these functions.

Fixes: https://github.com/rstudio/gt/issues/775
Fixes: https://github.com/rstudio/gt/issues/800
Fixes: https://github.com/rstudio/gt/issues/612

### CODE REVIEW

File: R/format_data.R [hunk 1]
```diff
@@ -1449,11 +1451,12 @@ fmt_date <- function(data,
     !column_classes_are_valid(
       data = data,
       columns = {{ columns }},
-      valid_classes = c("Date", "character")
+      valid_classes = c("Date", "POSIXct", "character")
```
Comment (jcheng5) [inline R/format_data.R:1454]: Change `POSIXct` to `POSIXt`

File: R/format_data.R [hunk 2]
```diff
@@ -1467,24 +1470,26 @@ fmt_date <- function(data,
     fns = list(
       default = function(x) {
 
-        # If `x` is of the `Date` type, simply make
-        # that a character vector
-        if (inherits(x, "Date")) {
+        # If `x` is of the `Date` or `POSIXct` types, simply transform
+        # `x` into a character vector
+        if (inherits(x, "Date") || inherits(x, "POSIXct")) {
           x <- as.character(x)
```
Comment (jcheng5) [inline R/format_data.R:1476]: I wouldn't do this conversion, just let strftime do it.

File: R/format_data.R [hunk 3]
```diff
@@ -1467,24 +1470,26 @@ fmt_date <- function(data,
...
         }
 
-        date <-
-          ifelse(grepl("^[0-9]*?\\:[0-9]*?", x), paste("1970-01-01", x), x) %>%
-          strftime(format = date_format_str)
+        # If the incoming string is a time then append the `1970-01-01` date
+        date <- ifelse(grepl("^[0-9]*?\\:[0-9]*?", x), paste("1970-01-01", x), x)
+
+        # Format the date string using `strftime()`
```
Comment (jcheng5) [inline R/format_data.R:1482]:
```
date <- tryCatch(as.POSIXlt(date), error = function(e) {
  # throw a friendly error that date(s) were invalid
})
```

File: R/format_data.R [hunk 4]
```diff
@@ -1467,24 +1470,26 @@ fmt_date <- function(data,
...
+        date <- strftime(date, format = date_format_str)
 
-        if (date_style %in% 2:12) {
-          date <- date %>% tidy_gsub(., "^0", "")
+        # Perform several cosmetic changes to the formatted date
+        if (date_format_str != "%F") {
```
Comment (jcheng5) [inline R/format_data.R:1486]: If you accept custom strftime format strings, then don't do this

File: R/format_data.R (hunk 3 above)
Comment (rich-iannone) [reply R/format_data.R:1486]: Now done.

File: R/format_data.R (hunk 2 above)
Comment (rich-iannone) [reply R/format_data.R]: Done.

File: R/format_data.R (hunk 1 above)
Comment (rich-iannone) [reply R/format_data.R]: Done.

File: R/format_data.R (hunk 4 above)
Comment (rich-iannone) [reply R/format_data.R:1490]: This codepath is avoided when providing a `format` in `fmt_datetime()`.

File: R/utils.R [hunk 5]
```diff
@@ -133,6 +133,18 @@ get_time_format <- function(time_style) {
...
+#' Are string values 24 hour times?
+#'
+#' Determine whether string values are representative of ISO 8601 time parts
+#' (in 24 hour time). Valid strings can be in the folowing formats: `hh::mm`,
+#' `hh::mm:ss`, and `hh::mm:ss.sss`.
+#'
+#' @noRd
+is_string_time <- function(x) {
+
+  is.character(x) & grepl("^\\d{2}:\\d{2}(:\\d{2}(\\.\\d+)?)?$", x)
```
Comment (jcheng5) [inline R/utils.R:145]: Let's make the first `{2}` a `{1,2}` instead

File: R/format_data.R [hunk 6]
```diff
@@ -1704,9 +1704,69 @@ fmt_time <- function(data,
...
+#' - `"%%"` -> `"%"` (the literal "`%`" character, in case you need it)
+#'
 #' @inheritParams fmt_number
 #' @inheritParams fmt_date
 #' @inheritParams fmt_time
+#' @param format An optional format code used for generating custom dates/times.
+#'   If used then the arguments governing preset styles (`date_style` and
+#'   `time_style`) will be ignored in favor of formatting via the `format`
+#'   string.
+#' @param tz Specification of the time zone and only used internally when
```
Comment (jcheng5) [inline R/format_data.R:1765]: Change "Specification of the time zone" to make it clear this is talking about the time zone of the _output_, not the input. To specify the input time zone, use the `tz` argument of `POSIXct`.

File: R/format_data.R [hunk 7]
```diff
@@ -1774,19 +1844,45 @@ fmt_datetime <- function(data,
...
+                call. = FALSE
+              )
+            }
+          )
 
-        # Format the date string using `strftime()`
-        date <- strftime(date, format = date_format_str)
+        # Format `datetime` into a date string using `strftime()` with
+        # the resolved formatting string
+        date <- strftime(datetime, format = date_format_str)
```
Comment (jcheng5) [inline R/format_data.R:1885]: This should also take `tz = tz`

File: R/format_data.R [hunk 8]
```diff
@@ -1799,11 +1895,9 @@ fmt_datetime <- function(data,
...
         #
 
-        # If the incoming string is a time then append the `1970-01-01` date
-        time <- ifelse(grepl("^[0-9]*?\\:[0-9]*?", x), paste("1970-01-01", x), x)
-
-        # Format the date string using `strftime()`
-        time <- strftime(time, format = time_format_str)
+        # Format `datetime` into a time string using `strftime()` with
+        # the resolved formatting string
+        time <- strftime(datetime, format = time_format_str)
```
Comment (jcheng5) [inline R/format_data.R:1900]: This should also take `tz = tz`

File: R/format_data.R [hunk 9]
```diff
@@ -1814,9 +1908,7 @@ fmt_datetime <- function(data,
           time <- gsub("am$", "AM", time)
         }
 
-        datetime <-
-          paste(date, time) %>%
-          tidy_gsub("NA NA", "NA")
+        datetime <- gsub("NA NA", "NA", paste(date, time))
```
Comment (jcheng5) [inline R/format_data.R:1911]: Let's create a helper/decorator function that prevents our formatter function from ever seeing NA inputs

File: tests/testthat/test-fmt_date_time.R [hunk 10]
```diff
@@ -539,4 +538,79 @@ test_that("the `fmt_datetime()` function works correctly", {
...
+        "August 05, 2017 09:45:23", "October 23, 2017 01:32:00",
+        "January 01, 2000 00:00:00"
+      )
+    )
+
+    expect_equal(
+      (tab %>%
+         fmt_datetime(
+           columns = "datetime",
+           format = "%B %d, %Y %H:%M:%S (%z)", tz = "GMT"
```
Comment (jcheng5) [inline tests/testthat/test-fmt_date_time.R:588]: Have tests that do different timezones as well

File: tests/testthat/test-fmt_date_time.R [hunk 11]
```diff
@@ -539,4 +538,79 @@ test_that("the `fmt_datetime()` function works correctly", {
...
+
+  #
+  # Format `datetime` using a custom `format` and verify the output
+  #
+
+  for (tab in gt_tables) {
+
+    expect_equal(
+      (tab %>%
+         fmt_datetime(
```
Comment (jcheng5) [inline tests/testthat/test-fmt_date_time.R:550]: Add a test where a timezone is applied to the input (`datetime` column)

**Review by krlmlr — COMMENTED** [review]
I had a very quick look. My biggest question here is: why does gt need to be concerned with parsing date/time/datetime values from strings? How bad would it be if gt accepted only properly typed data? Perhaps we could support two code paths, one for character input (legacy, using existing code) and one for typed input (without support for parsing)?

File: R/utils.R [hunk 12]
```diff
@@ -40,48 +40,206 @@ time_formats <- function() {
...
+}
+
+is_basic_8601_dt <- function(x) {
+
+  grepl("^[0-9]{8}T([0-9]{6}(\\.[0-9]+)?|[0-9]{4}|[0-9]{2})(|Z|(\\+|-)[0-9]{2}([:]?[0-9]{2})?)?$", x)
+}
+
+has_dt_tz <- function(x) {
+
+  grepl(".+(Z|z|(\\+|-)[0-9]{2}([:]?[0-9]{2})?)?$", "2020-04-21T13:04:58.321Z")
```
[the remark this answers is not in the document]
Comment (rich-iannone) [reply R/utils.R]: Now corrected. Currently it isn't used (should be removed before merging if never used).

File: R/utils.R (hunk 5 above)
Comment (rich-iannone) [reply R/utils.R]: Now fixed.

File: R/format_data.R (hunk 6 above)
Comment (rich-iannone) [reply R/format_data.R]: Now changed to convey that this is the output time zone. Now reads:

```
The time zone for printing dates/times (i.e., the output). The
default of `"UTC"` will render dates/times in UTC (and this matches the
**lubridate** package default for parsing date-times). If providing a
different time zone, it must be one that is recognized by the user's
operating system (a vector of all valid `tz` values can be produced with
[OlsonNames()]).
```

File: R/format_data.R (hunk 7 above)
Comment (rich-iannone) [reply R/format_data.R]: Done.

File: R/format_data.R (hunk 8 above)
Comment (rich-iannone) [reply R/format_data.R]: Done.

File: tests/testthat/test-fmt_date_time.R (hunk 10 above)
Comment (rich-iannone) [reply tests/testthat/test-fmt_date_time.R:588]: Now done.

File: R/format_data.R [hunk 13]
```diff
@@ -1602,21 +1619,44 @@ fmt_time <- function(data,
...
+        # Format the date string using `strftime()`
+        time <- strftime(time, format = time_format_str)
+
+        # Perform several cosmetic changes to the formatted time
+        if (grepl("%P$", time_format_str)) {
+
+          time <- gsub("^0", "", time)
+          time <- gsub(" 0([0-9])", " \\1", time)
+          time <- gsub("pm$", "PM", time)
+          time <- gsub("am$", "AM", time)
```
Comment (jcheng5) [inline R/format_data.R:1656]: Can we add word boundaries around "am" and "pm"?

File: R/format_data.R [hunk 14]
```diff
@@ -1702,29 +1809,42 @@ fmt_datetime <- function(data,
                          columns,
                          rows = everything(),
                          date_style = 2,
-                         time_style = 2) {
+                         time_style = 2,
+                         sep = " ",
+                         format = NULL,
+                         tz = "UTC",
```
Comment (jcheng5) [inline R/format_data.R:1815]: Default to `tz = NULL`

File: R/format_data.R [hunk 15]
```diff
@@ -1733,42 +1853,72 @@ fmt_datetime <- function(data,
...
+        #
 
-        time <-
-          ifelse(grepl("^[0-9]*?\\:[0-9]*?", x), paste("1970-01-01", x), x) %>%
-          strftime(format = time_format_str)
+        # Convert incoming values to POSIXlt but provide a friendly error
+        # if the values cannot be parsed by `as.POSIXlt()`
+        datetime <-
+          tryCatch(
+            as.POSIXlt(x, tz = tz),
```
Comment (jcheng5) [inline R/format_data.R:1884]: Don't pass `tz = tz` here

File: R/format_data.R [hunk 16]
```diff
@@ -1702,29 +1809,42 @@ fmt_datetime <- function(data,
...
     !column_classes_are_valid(
       data = data,
       columns = {{ columns }},
-      valid_classes = "character"
-    )) {
-    stop(
-      "The `fmt_datetime()` function can only be used on `columns` with `character` values.",
-      call. = FALSE
-    )
+      valid_classes = c("Date", "POSIXt", "character"))
```
Comment (jcheng5) [inline R/format_data.R:1841]: If we don't accept `POSIXlt` as input, then valid class should be `"POSIXct"`

File: R/format_data.R [hunk 17]
```diff
@@ -1733,42 +1853,72 @@ fmt_datetime <- function(data,
...
+        # the resolved formatting string
+        time <- strftime(datetime, format = time_format_str, tz = tz)
+
+        # Perform several cosmetic changes to the formatted time
+        if (grepl("%P$", time_format_str)) {
+
+          time <- gsub("^0", "", time)
+          time <- gsub(" 0([0-9])", " \\1", time)
+          time <- gsub("pm$", "PM", time)
+          time <- gsub("am$", "AM", time)
```
Comment (jcheng5) [inline R/format_data.R:1918]: Same as above

File: R/utils_formatters.R [hunk 18]
```diff
@@ -740,11 +754,11 @@ num_fmt_factory <- function(context,
     x_str_vals <-
       x_vals %>%
       # Format all non-NA x values with a formatting function
-      format_fn(context) %>%
+      format_fn(context = context) %>%
       # If in a LaTeX context, wrap values in math mode
-      to_latex_math_mode(context) %>%
+      to_latex_math_mode(context = context, active = use_latex_math_mode) %>%
```
Comment (jcheng5) [inline R/utils_formatters.R:759]: Use a conditional here inside of `{...}` instead of passing a flag

File: R/format_data.R (hunk 14 above)
Comment (rich-iannone) [reply R/format_data.R:1815]: Now changed.

File: R/format_data.R (hunk 15 above)
Comment (rich-iannone) [reply R/format_data.R:1884]: Now removed.

File: R/format_data.R (hunk 16 above)
Comment (rich-iannone) [reply R/format_data.R:1841]: Now changed.

File: R/format_data.R (hunk 13 above)
Comment (rich-iannone) [reply R/format_data.R:1656]: Now changed.

File: R/format_data.R (hunk 17 above)
Comment (rich-iannone) [reply R/format_data.R:1918]: Fixed.

File: R/utils_formatters.R (hunk 18 above)
Comment (rich-iannone) [reply R/utils_formatters.R]: Now done.

**Review by jcheng5 — APPROVED** [review]

**Pull request merged**